lib/common/tests/xml/pcmk__xe_foreach_child_test.c

```diff
@@ -0,0 +1,148 @@
+    function_called();
+    assert_string_equal((char *) userdata, crm_element_name(xml));
+}
+
+const char *str1 =
+    "<xml>\n"
+    "  <level1>\n"
+    "  </level1>\n"
+    "  <level1>\n"
+    "  </level1>\n"
+    "  <level1>\n"
+    "  </level1>\n"
+    "</xml>";
+
+static void
+bad_input(void **state) {
+    xmlNode *xml = string2xml(str1);
+
+    pcmk__assert_asserts(pcmk__xe_foreach_child(NULL, NULL, compare_name_handler, NULL));
+    pcmk__assert_asserts(pcmk__xe_foreach_child(xml, NULL, NULL, NULL));
+
+    free_xml(xml);
+}
+
+static void
+name_given_test(void **state) {
+    xmlNode *xml = string2xml(str1);
+
+    /* The handler should be called once for every <level1> node. */
+    expect_function_call(compare_name_handler);
+    expect_function_call(compare_name_handler);
+    expect_function_call(compare_name_handler);
+
+    pcmk__xe_foreach_child(xml, "level1", compare_name_handler, (void *) "level1");
+    free_xml(xml);
+}
+
+static void
+no_name_given_test(void **state) {
+    xmlNode *xml = string2xml(str1);
+
+    /* The handler should be called once for every <level1> node. */
+    expect_function_call(compare_name_handler);
+    expect_function_call(compare_name_handler);
+    expect_function_call(compare_name_handler);
+
+    pcmk__xe_foreach_child(xml, NULL, compare_name_handler, (void *) "level1");
+    free_xml(xml);
+}
+
+static void
+name_doesnt_exist_test(void **state) {
+    xmlNode *xml = string2xml(str1);
+    pcmk__xe_foreach_child(xml, "xxx", compare_name_handler, NULL);
+    free_xml(xml);
+}
+
+const char *str2 =
+    "<xml>\n"
+    "  <level1>\n"
+    "    <level2>\n"
+    "    </level2>\n"
+    "  </level1>\n"
+    "  <level1>\n"
+    "    <level2>\n"
+    "      <level3>\n"
+    "      </level3>\n"
+    "    </level2>\n"
+    "    <level2>\n"
+    "    </level2>\n"
+    "  </level1>\n"
+    "</xml>";
+
+static void
+multiple_levels_test(void **state) {
+    xmlNode *xml = string2xml(str2);
+
+    /* The handler should be called once for every <level1> node. */
+    expect_function_call(compare_name_handler);
+    expect_function_call(compare_name_handler);
+
+    pcmk__xe_foreach_child(xml, "level1", compare_name_handler, (void *) "level1");
+    free_xml(xml);
+}
+
+static void
+multiple_levels_no_name_test(void **state) {
+    xmlNode *xml = string2xml(str2);
+
+    /* The handler should be called once for every <level1> node. */
+    expect_function_call(compare_name_handler);
+    expect_function_call(compare_name_handler);
+
+    pcmk__xe_foreach_child(xml, NULL, compare_name_handler, (void *) "level1");
+    free_xml(xml);
+}
+
+const char *str3 =
+    "<xml>\n"
+    "  <node1>\n"
+    "  </node1>\n"
+    "  <node2>\n"
+    "  </node2>\n"
+    "  <node3>\n"
+    "  </node3>\n"
+    "</xml>";
+
+static void any_of_handler(xmlNode *xml, void *userdata) {
+    function_called();
+    assert_true(pcmk__str_any_of(crm_element_name(xml), "node1", "node2", "node3", NULL));
+}
```

Comment:
one more test would be worthwhile to ensure we pull out one of these nodes correctly

maybe even a test for each nodeN so we're checking first, middle, last

Reply:
Added.